experiments/burrow/burrow_module.py

```diff
@@ -0,0 +1,285 @@
+        self.deployed_contract = None
+        self.validator_addresses = {}
+        self.experiment.message_callback = self
+
+        self.submitted_transactions = {}
+        self.confirmed_transactions = {}
+
+    def on_all_vars_received(self):
+        super().on_all_vars_received()
+        self.transactions_manager.transfer = self.transfer
+
+    @experiment_callback
+    def transfer(self):
+        validator_peer_id = ((self.my_id - 1) % self.num_validators) + 1
+        host, _ = self.experiment.get_peer_ip_port_by_id(validator_peer_id)
+
+        def create_and_submit_tx():
+            url = 'http://%s:%d' % (host, 12000 + validator_peer_id)
+            w3 = Web3(Web3.HTTPProvider(url))
+            random_id = ''.join(random.choice(string.ascii_uppercase) for _ in range(5))
+            submit_time = int(round(time.time() * 1000))
+            self.submitted_transactions[random_id] = submit_time
+
+            tx_hash = self.deployed_contract.functions.transfer(
+                w3.toChecksumAddress("EB77C5D07D50D9853EF90CB6B32E38755A9BDF2F"), 1, random_id).transact(
+                {"from": w3.toChecksumAddress(self.validator_addresses[validator_peer_id])})
+            try:
+                _ = w3.eth.waitForTransactionReceipt(tx_hash)
+                confirm_time = int(round(time.time() * 1000))
+                self.confirmed_transactions[random_id] = confirm_time
+            except TimeExhausted:
+                print("Time exhausted for tx with hash: %d" % tx_hash)
+
+        t = Thread(target=create_and_submit_tx)
+        t.daemon = True
+        t.start()
+
+    def on_message(self, from_id, msg_type, msg):
+        self._logger.info("Received message with type %s from peer %d", msg_type, from_id)
+        if msg_type == b"validator_address":
+            validator_address = msg.decode()
+            self.validator_addresses[from_id] = validator_address
+        elif msg_type == b"contract_address":
+            self.deployed_contract_address = msg.decode()
+        elif msg_type == b"contract_abi":
+            self.deployed_contract_abi = json.loads(unhexlify(msg).decode())
+
+        if self.deployed_contract_address and self.deployed_contract_abi:
+            validator_peer_id = ((self.my_id - 1) % self.num_validators) + 1
+            host, _ = self.experiment.get_peer_ip_port_by_id(validator_peer_id)
+            url = 'http://%s:%d' % (host, 12000 + validator_peer_id)
+            w3 = Web3(Web3.HTTPProvider(url))
+            self.deployed_contract = w3.eth.contract(address=self.deployed_contract_address,
+                                                     abi=self.deployed_contract_abi)
+
+    @experiment_callback
+    def generate_config(self):
+        """
+        Generate the initial configuration files.
+        """
+        self._logger.info("Generating Burrow config...")
+
+        # Remove old config directory
+        shutil.rmtree("/home/martijn/burrow_data", ignore_errors=True)
```

Comment:
It seems to me like all of these scripts are heavily tied to your ecosystem. At this moment, it seems this code serves to show others what you are running and not to allow others to run what you are running.

For reuse, I'd suggest (1) adding a (short) description to each file to document what software needs to be installed. I'd also suggest (2) to refactor of all the paths that currently point to your personal directory into a variable (i.e., `self.install_dir / "burrow_data"` instead of `"/home/martijn/burrow_data"`).

I'll leave this as a suggestion and not a request for changes. After all, I'm not sure whether you even want others to run this code (on the other hand, why push this code to the shared repo then?). Let me know if this is out of scope and I'll approve the PR as needed.

Reply:
Thanks for the suggestions! Indeed, the current code is tightly integrated with our infrastructure but that has always been the case, even with the DAS5 code. Even though I acknowledge it is not a good pattern, my goal is to make Gumby suitable to run these blockchain experiments on our (Surfnet) infrastructure.

My OP might have been a bit misleading then since it does not support running these experiments on arbitrary servers without a specific setup. Many of these blockchain systems require multiple binaries/support for different compilers etc. For now, I try to keep it simple but ideally, we want to automate the setup phase. To some extent, this has been done already (e.g., the generation of genesis files).

My goal with this PR is to provide the experiment code since it gives valuable insights into the setup/operation of the tested systems and how to deploy a private testnet, even though I acknowledge that it is very, very hard to run it in a different environment.

BTW, we can probably move some of these variables to the configuration. However, I would first like to rework the configuration system, so I'm delaying that effort for now.